Objects/genericaliasobject.c

```diff
@@ -626,6 +639,50 @@ static PyNumberMethods ga_as_number = {
         .nb_or = _Py_union_type_or, // Add __or__ function
 };
 
+static PyObject *
+ga_iternext(gaiterobject *gi) {
+    if (gi->obj == NULL) {
+        return NULL;
+    }
+    gaobject *alias = (gaobject *) gi->obj;
+    PyObject *starred_tuple = Py_GenericAlias(alias->origin, alias->args);
+    ((gaobject * ) starred_tuple)->starred = 1;
+    Py_SETREF(gi->obj, NULL);
```

Comment:
> "Before fields which refer to other containers are invalidated, PyObject_GC_UnTrack() must be called"

The line preceding that says `Similarly, the deallocator for the object must conform to a similar pair of rules:`. In this case, the "deallocator" refers to the deallocator function ([tp_dealloc](https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_dealloc)). Not the refcounting mechanism.

Hopefully I didn't misunderstand, and maybe you know this already, but at the risk of being a nag:
CPython has two distinct (and largely separate) automatic memory management mechanisms, reference counting, and the GC:
1. Refcounting -- pretty simple, once a refcount hits zero, `tp_dealloc` is called and the object is deallocated. This is checked in every `Py_DECREF`.
2. GC -- this runs periodically based on certain thresholds. We only need a GC due to circular references (e.g., a class's `__mro__` contains a reference to itself).

(More detail here https://devguide.python.org/garbage_collector/)

While the `Py_DECREF` function can trigger the GC, we're mostly still in the refcounting realm here. Decrefing here should be correct (you can see implementations of other iterators, like list iter's iternext or rang iter's iternext). The only times you have to care about GC is when you're dealing with cyclic objects and `tp_clear`, tp_traverse`, `tp_init` `tp_dealloc`, and `tp_free`. Outside of those functions, you can just ignore them (for the most part).

PS. Sorry if there are inaccuracies anywhere here. Honestly my GC knowledge is rusty and Pablo is definitely a better source of info for this than I am.

Reply:
Wait :facepalm:. I think I get what you're asking now. Sorry about the long ramble. But yes you should still be decrefing here. Once the iterator exhausts itself it shouldn't hold on to the underlying sequence anymore. Otherwise, if some code holds a reference to the exhausted iterator, the underlying sequence will never be freed.

Reply:
Ok, fair enough :)